File: ruby/custom-set/custom_set.rb
=begin
Write your code for the 'Custom Set' exercise in this file. Make the tests in
`custom_set_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/custom-set` directory.
=end

class CustomSet
  def initialize(ary)
    @ary = ary.uniq
  end

  def empty?
    @ary.empty?
  end

  def member?(elem)
    @ary.member?(elem)
  end

  def subset?(set)
    @ary.size.zero? || @ary.all? { set.member? _1 }
  end

  def disjoint?(set)
    @ary.none? { set.member? _1 }
  end

  def add(elem)
    @ary << elem
    @ary.uniq!
    @ary
  end

  def size
    @ary.size
  end

  def difference(set)
    @ary.reject { set.member?(_1) }
  end

  def intersection(set)
    @ary.filter { set.member?(_1) }
  end

  def union(set)
    @ary += set.difference(self)
  end

  def ==(other)
    @ary.size == other.size && @ary.all? { other.member? _1 }
  end
end

File: csharp/custom-set/CustomSet.cs
using System.Linq;

public class CustomSet(params int[] values)
{
    private readonly int[] _values = values.Distinct().ToArray();

    public CustomSet Add(int value) =>
        !Contains(value) ? new(_values.Append(value).ToArray()) : new (_values);

    public bool Empty() => _values.Length == 0;
    public bool Contains(int value) => _values.Contains(value);
    public bool Subset(CustomSet right) => _values.All(right.Contains);
    public bool Disjoint(CustomSet right) => _values.All(x => !right.Contains(x));
    public CustomSet Intersection(CustomSet right) => new(_values.Where(right.Contains).ToArray());
    public CustomSet Difference(CustomSet right) => new(_values.Where(x => !right.Contains(x)).ToArray());
    public CustomSet Union(CustomSet right) => new(_values.Union(right._values).ToArray());
    public override bool Equals(object obj) => obj is CustomSet set && Equals(set);
    public bool Equals(CustomSet right) => _values.All(right.Contains) && right._values.All(Contains);
    public override int GetHashCode() => _values.GetHashCode();
}
